### PR TITLE
162 generator add the probe generator

### PR DESCRIPTION
**Pull request by kmariuszk**

### CODE REVIEW

File: docs/src/explanation/generators/probe.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+```
+
+# `ProbeGenerator`
+
+The `ProbeGenerator` is designed to navigate the trade-offs between costs and robustness in algorithmic recourse.
+
+## Description
+
+The goal of ProbeGenerator is to find a recourse x' whose prediction at any point y within some set around x' belongs to the positive class with probability 1 - r, where r is the recourse invalidation rate. It minimizes the gap between the achieved and desired recourse invalidation rates, minimizes recourse costs, and also ensures that the resulting recourse achieves a positive model prediction.
+Usage
```
Comment (A-Salmon) [inline docs/src/explanation/generators/probe.md:12]: Small nitpick, I don't think the word "usage" is supposed to be here

File: src/counterfactuals/functions.jl
Comment (A-Salmon) [inline src/counterfactuals/functions.jl:1]: Make sure the documentation matches the methods! There's several parameters present in the method that aren't found in the documentation (see: CounterfactualExplanation constructor). Other than that, could you clarify on the change in indentation a bit? Is this to follow a convention or is there some other reason?

**Review by A-Salmon — CHANGES_REQUESTED** [review]
In general, I think this looks great, nice work! I do have a couple things to note about it, mainly related to documentation. Please address these.

File: src/generators/gradient_based/functions.jl [hunk 2]
```diff
@@ -30,7 +37,8 @@ end
 """
     ∇(generator::AbstractGradientBasedGenerator, M::Models.AbstractDifferentiableModel, ce::AbstractCounterfactualExplanation)
 
-The default method to compute the gradient of the counterfactual search objective for gradient-based generators. It simply computes the weighted sum over partial derivates. It assumes that `Zygote.jl` has gradient access.
+The default method to compute the gradient of the counterfactual search objective for gradient-based generators.
+It simply computes the weighted sum over partial derivates. It assumes that `Zygote.jl` has gradient access.
 """
 function ∇(
```
Comment (RaunoArike) [inline src/generators/gradient_based/functions.jl:43]: Small nitpick, but I think that it would be clearer if the hinge_loss term was added inside this method rather than inside ∂ℓ. This makes it easier to see that Probe's loss function has three terms - otherwise, the robustness term seems more like a component of the model's loss function.

File: src/generators/gradient_based/probe.jl [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+Calculate the invalidation rate of a counterfactual explanation.
+
+# Arguments
+- `ce::AbstractCounterfactualExplanation`: The counterfactual explanation to calculate the invalidation rate for.
+- `kwargs`: Additional keyword arguments to pass to the function.
+
+# Returns
+The invalidation rate of the counterfactual explanation.
+"""
+function invalidation_rate(ce::AbstractCounterfactualExplanation; kwargs...)
```
Comment (RaunoArike) [inline src/generators/gradient_based/probe.jl:44]: Since the kwargs don't seem to be used anywhere inside the function body, is it necessary to have them in the parameter list?

File: src/generators/gradient_based/probe.jl [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    f_loss = logits(ce.M, CounterfactualExplanations.decode_state(ce))[ce.target]
+    grad = []
+    # This has to be done with a for loop because flux does not know how to take a gradient from an array of logits.
```
Comment (RaunoArike) [inline src/generators/gradient_based/probe.jl:47]: I think Flux is supposed to be able to take a gradient from an array of logits. I remember there being a valid reason for why a for-loop is needed here, but is it really that Flux just cannot take gradients with respect to an array at all?

**Review by RaunoArike — CHANGES_REQUESTED** [review]
Good job overall, but I think both me and Simon had some questions that should be resolved before giving a final approval.

File: src/counterfactuals/functions.jl
Comment (laurikskl) [reply src/counterfactuals/functions.jl:1]: Great point on the constructor.

For indentation -  we haven't changed anything manually, it's just the formatter that did that I think?

File: src/generators/gradient_based/probe.jl (hunk 3 above)
Comment (laurikskl) [reply src/generators/gradient_based/probe.jl]: I've removed it

File: src/generators/gradient_based/functions.jl (hunk 2 above)
Comment (laurikskl) [reply src/generators/gradient_based/functions.jl:40]: Changed it to your liking, I do agree that this is a bit cleaner.

File: src/generators/Generators.jl
[the remark this answers is not in the document]
Comment (laurikskl) [reply src/generators/Generators.jl:1]: There is no difference, hingeLoss does not exist anymore.

**Review by RaunoArike — APPROVED** [review]
Thanks for making the requested changes, I think this PR is ready to be merged now.

**Review by A-Salmon — APPROVED** [review]
